### ISSUE TITLE
Report "value is always false/true" when HasValue check is unnecessary

### ISSUE THREAD

**Issue by cston** [dotnet/roslyn#31516]
Report `HDN_NullCheckIsProbablyAlwaysFalse/True` for `HasValue` checks below:
```C#
class Program
{
    static void F<T>(T? t) where T : struct
    {
        if (!t.HasValue) return;

        if (!t.HasValue) // always false
            _ = t.Value;

        if (t.HasValue) // always true
            _ = t.Value;
    }
}
```

**Comment by alrz** [comment]
Does this relate to https://github.com/dotnet/roslyn/issues/25592 (a general warning) or it's just a special case?

**Comment by cston** [comment]
@alrz, `HasValue` checks are a special case where we could report "always true/false" diagnostics. Also, the "always true/false" diagnostics that are reported currently (for reference types) are hidden diagnostics rather than warnings. It looks like #25592 is specifically about reporting warnings.

**Comment by gafter** [comment]
Today the LDM confirmed our intent to drop the HDN diagnostics.
